Functionality when payment on paytomorrow succeeds/cancels

abhishekgupta5 opened the pull request:
**Issue:** https://linear.app/nebulab-retainers/issue/ABU-9/create-a-controller-to-handle-when-application-is-successfully-created
and
https://linear.app/nebulab-retainers/issue/ABU-17/create-a-controller-to-handle-when-application-is-cancel

- [x] Template controller to handle success/cancel
- [x] Controller specs

This PR depends on https://github.com/nebulab/solidus_pay_tomorrow/pull/12

DanielePalombo committed:
Add base_url configuration

1. Added base_url to the PayTomorrow configuration
2. Added Spree::TestingSupport::UrlHelpers to be used in specs
3. Used 1 & 2 in constructing returnUrl, cancelURL in existing specs

abhishekgupta5 commented:
Force-pushed because had to rebase from master

DanielePalombo commented on spec/services/solidus_pay_tomorrow/client/create_order_service_spec.rb:
```diff
@@ -39,9 +39,9 @@ def expected_body
-        returnUrl: 'https://domain.com/webhooks/return',
-        cancelUrl: 'https://domain.com/webhooks/cancel',
-        notifyUrl: 'https://domain.com/webhooks/notify',
+        returnUrl: "#{SolidusPayTomorrow.config.base_url}/return",
+        cancelUrl: "#{SolidusPayTomorrow.config.base_url}/cancel",
+        notifyUrl: "#{SolidusPayTomorrow.config.base_url}/notify",
```
Considering that `return` and `cancel` have been implemented, what do you think about using the route helpers `pay_tomorrow_return_path` and `pay_tomorrow_cancel_path`?

(They need to be scoped under Spree::Core::Engine.routes.url_helpers )

abhishekgupta5 replied:
The problem with this is ->
`pay_tomorrow_return_path` = `/pay_tomorrow/return`
&
`SolidusPayTomorrow.config.base_url` ~ `https://80a5-125-63-97-62.in.ngrok.io/pay_tomorrow`

Hence if we'll try to construct returnUrl,
"#{SolidusPayTomorrow.config.base_url}#{spree.pay_tomorrow_return_path}"

it'll be 
`https://80a5-125-63-97-62.in.ngrok.io/pay_tomorrow/pay_tomorrow/return`

The path would be duplicated so we'll have to do some hackish business in order to fix it.
Doable with URI.parse but not sure if it is worth it.

Wdyt?

DanielePalombo replied:
The value of `SolidusPayTomorrow.config.base_url` is decided by us. 
We can change it to `https://80a5-125-63-97-62.in.ngrok.io/` without any issue.

abhishekgupta5 replied:
Cool then, we'll change the ENV value and use `#{SolidusPayTomorrow.config.base_url}#{spree.pay_tomorrow_return_path}`

abhishekgupta5 replied:
Added this in https://github.com/nebulab/solidus_pay_tomorrow/pull/11/commits/cd13268ce6ff1f156e33617bbb8ec22de3a85167

DanielePalombo commented on lib/generators/solidus_pay_tomorrow/install/templates/solidus_pay_tomorrow_order_application_controller.rb:
```diff
@@ -0,0 +1,24 @@
+    def success
+      # FIXME: will have to find the correct payment object properly.
+      # This is error prone
+      payment = current_order.payments.where(state: :checkout).last
```
As we already discussed, we have to link this payment with the PaymentSource when the payment is created, so we can filter the payments by the `source`.

abhishekgupta5 replied:
Added PaymentSource in [this PR](https://github.com/nebulab/solidus_pay_tomorrow/pull/12) and fixed the query here in [this commit](https://github.com/nebulab/solidus_pay_tomorrow/pull/11/commits/eff5820dc37118df52146515590a7da2392260cc)

DanielePalombo commented on config/initializers/solidus_pay_tomorrow.rb:
```diff
@@ -0,0 +1,5 @@
+# frozen_string_literal: true
+
+SolidusPayTomorrow.configure do |config|
+  config.base_url = ENV['PAY_TOMORROW_REDIRECT_BASE_URL']
```
What do you think about adding the following command at the end of the sandbox script?

`sed -i '' "/Rails.application.configure do/{p;s/.*/  config.hosts << URI.parse(ENV['PAY_TOMORROW_REDIRECT_BASE_URL']).hostname/;}" config/environments/development.rb # rubocop:disable Layout/LineLength`

In this way, all the specified URL will be accepted automatically as incoming connections.

abhishekgupta5 replied:
Good idea! Will save the ngrok step for the dev.
But then, I think we'd also add -
`export PAY_TOMORROW_REDIRECT_BASE_URL = ngrok-url` sometime before the sed command.

Because if the ENV variable is not set, then the sed command will fail. Now the question is, we'd also have the ngrok-URL before building the sandbox.

Thoughts?

DanielePalombo replied:
Probably we can change it to:

```
if [ -z $PAY_TOMORROW_REDIRECT_BASE_URL]
then
sed -i '' "/Rails.application.configure do/{p;s/.*/  config.hosts << URI.parse(ENV['PAY_TOMORROW_REDIRECT_BASE_URL']).hostname/;}" config/environments/development.rb # rubocop:disable Layout/LineLength
fi
```

In this way, the sed command will be executed only when the variable is set

abhishekgupta5 replied:
Added in https://github.com/nebulab/solidus_pay_tomorrow/pull/11/commits/6486d40f7c5ddb4f08e98e140af0587406947c29

DanielePalombo reviewed (changes requested):
If the payment is rejected by paytomorrow the user is redirected to pay tomorrow site. Is there anything that we can set to prevent it and redirect the user to the cancel URL?

abhishekgupta5 committed:
Added SolidusPayTomorrowPaymentSource

1. Added the migration for table
2. Fixed table_name_prefix so that the model can refer to the table
3. Added model factory and specs

abhishekgupta5 committed:
Create PaymentSource along with payment

Since payment's source is ought to be present,
create it before creating the payment object

Fixed failing specs

DanielePalombo reviewed (changes requested):
This need to be rebased on top of ABU-18, right?

abhishekgupta5 commented:
> This need to be rebased on top of ABU-18, right?

Yes

abhishekgupta5 committed:
Updated bin/sandbox

Adds $PAY_TOMORROW_REDIRECT_BASE_URL to config.hosts if present.
This variable is required for paytomorrow redirects.

In dev, this var holds the ngrok url and the step to add it
to config.hosts is automated with the sed command

If the variable is not present, info to set it is logged

abhishekgupta5 committed:
Add template controller for success/cancel

    Added the template controller with this functionality -
    We initiate the payment in :checkout state before this step

    Success scenario
    - When payment on PayTomorrow succeeds, mark the payment as
      :pending
    - Redirect to the next checkout(Confirm) step

    Failed scenario
    - When payment on PayTomorrow is cancelled, mark the payment
      as :invalid
    - Redirect to the same checkout(payment) step

    * Mount SolidusPayTomorrow route in install generator

abhishekgupta5 committed:
Added and configured shoulda_matchers

abhishekgupta5 commented:
> If the payment is rejected by paytomorrow the user is redirected to pay tomorrow site. Is there anything that we can set to prevent it and redirect the user to the cancel URL?

I'll reproduce this and see what we can do. But I think it should be a concern for a separate PR
